### PR TITLE
feat(logging): Allow configuring logging verbosity level

### PR DESCRIPTION
**Pull request by garyhodgson**
## Summary

Introduces logr library as per https://github.com/TwiN/gatus/discussions/871

I'm not a Go Developer, so apologies if the attempt isn't very go-ish.

I couldn't see a way to use enums for mapping the log level names to logr.Level ints, so I made use of this library: https://hermannm.dev/enumnames

I performed a few manual tests, not sure how unit tests of log output would work.

## Checklist
- [X] Tested and/or added tests to validate that the changes work as intended, if applicable.
- [X] Updated documentation in `README.md`, if applicable.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -47,13 +49,23 @@ var (
...
+	})
 )
 
 // Config is the main configuration structure
 type Config struct {
 	// Debug Whether to enable debug logs
 	Debug bool `yaml:"debug,omitempty"`
 
+	// Log level: DEBUG, INFO, WARN, ERROR.  Default: INFO
+	LogLevel string `yaml:"log-level,omitempty"`
```
Comment (TwiN) [inline config/config.go:67]: I just made a couple changes to the library - you should now be able to just use `logr.Level` as I've made it into a string.

This way, you won't have to worry about converting anything.

File: config/config.go [hunk 2]
```diff
@@ -204,6 +216,11 @@ func LoadConfiguration(configPath string) (*Config, error) {
 	}
 	config.configPath = usedConfigPath
 	config.UpdateLastFileModTime()
+
+	if config.LogLevel == "" {
+		config.LogLevel = "INFO"
+	}
```
Comment (TwiN) [inline config/config.go:222]:
```suggestion
	if config.LogLevel == "" {
		config.LogLevel = logr.LevelInfo
	}
```
once you took care of the previous suggestion, you should be able to just use this directly

File: main.go [hunk 3]
```diff
@@ -57,6 +61,19 @@ func save() {
...
+	if config.LogLevels.ContainsName(cfg.LogLevel) {
+		log.Printf("[main.configureLogging] Log Level is %s", cfg.LogLevel)
+		key, ok := config.LogLevels.GetKey(cfg.LogLevel)
+		if ok {
+			logr.SetThreshold(key)
+		}
+	} else {
+		log.Printf("[main.configureLogging] Unrecognised log level '%s', defaulting to INFO. Allowed values are [DEBUG, INFO, WARN, ERROR]", cfg.LogLevel)
+	}
+}
```
Comment (TwiN) [inline main.go:75]: I also just introduced the `logr.LevelFromString` function, which should help you with this.

The aforementioned function should return an error if the existing value isn't valid

**Review by TwiN — CHANGES_REQUESTED** [review]
Given that you've never written Go before, you're doing excellent work!

**Comment by TwiN** [comment]
FYI, don't hesitate to let me know if you want me to add something else to the logr library.

File: config/config.go (hunk 2 above)
Comment (TwiN) [reply config/config.go:222]: it's fine to have this anyways, but I also just made a change in the library to make it so if an invalid/unsupported threshold is set, it automatically defaults to INFO

**Comment by garyhodgson** [comment]
Hi. All good suggestions, the changes to logr make it a easier to integrate :+1: 

As far as suggestions, I can imagine convenience methods such as `isDebug()` for `logr.GetThreshold() == logr.LevelDebug` checks would prove useful.

File: watchdog/watchdog.go [hunk 4]
```diff
@@ -64,37 +66,36 @@ func execute(ep *endpoint.Endpoint, alertingConfig *alerting.Config, maintenance
...
-		log.Println("[watchdog.execute] No connectivity; skipping execution")
+		logr.Infof("[watchdog.execute] No connectivity; skipping execution")
 		return
 	}
-	if debug {
-		log.Printf("[watchdog.execute] Monitoring group=%s; endpoint=%s", ep.Group, ep.Name)
-	}
+
+	logr.Debugf("[watchdog.execute] Monitoring group=%s; endpoint=%s", ep.Group, ep.Name)
+
```
Comment (TwiN) [inline watchdog/watchdog.go:74]:
```suggestion
	logr.Debugf("[watchdog.execute] Monitoring group=%s; endpoint=%s", ep.Group, ep.Name)
```

File: main.go [hunk 5]
```diff
@@ -23,6 +25,8 @@ func main() {
 	if err != nil {
 		panic(err)
 	}
+
```
Comment (TwiN) [inline main.go:28]:
```suggestion
```

File: main.go [hunk 6]
```diff
@@ -12,6 +12,8 @@ import (
 	"github.com/TwiN/gatus/v5/controller"
 	"github.com/TwiN/gatus/v5/storage/store"
 	"github.com/TwiN/gatus/v5/watchdog"
+
```
Comment (TwiN) [inline main.go:15]:
```suggestion
```

File: go.mod [hunk 7]
```diff
@@ -1,13 +1,16 @@
 module github.com/TwiN/gatus/v5
 
-go 1.22.2
+go 1.22.4
+
+toolchain go1.23.0
```
Comment (TwiN) [inline go.mod:5]:
```suggestion
```

File: config/config.go [hunk 8]
```diff
@@ -54,6 +55,9 @@ type Config struct {
 	// Debug Whether to enable debug logs
 	Debug bool `yaml:"debug,omitempty"`
 
+	// Log level: DEBUG, INFO, WARN, ERROR.  Default: INFO
+	LogLevel logr.Level `yaml:"log-level,omitempty"`
```
Comment (TwiN) [inline config/config.go:59]:
```suggestion
	// LogLevel is one of DEBUG, INFO, WARN and ERROR. Defaults to INFO
	LogLevel logr.Level `yaml:"log-level,omitempty"`
```

File: watchdog/watchdog.go [hunk 9]
```diff
@@ -64,37 +66,36 @@ func execute(ep *endpoint.Endpoint, alertingConfig *alerting.Config, maintenance
...
-	} else if debug {
-		log.Println("[watchdog.execute] Not handling alerting because currently in the maintenance window")
-	}
-	if debug {
-		log.Printf("[watchdog.execute] Waiting for interval=%s before monitoring group=%s endpoint=%s again", ep.Interval, ep.Group, ep.Name)
+		HandleAlerting(ep, result, alertingConfig, logr.GetThreshold() == logr.LevelDebug)
+	} else {
+		logr.Debugf("[watchdog.execute] Not handling alerting because currently in the maintenance window")
 	}
+
```
Comment (TwiN) [inline watchdog/watchdog.go:91]:
```suggestion
```

**Review by TwiN — CHANGES_REQUESTED** [review]

**Comment by g-hodgson-tup** [comment]
Hi - requested changes have been pushed.

File: go.mod [hunk 10]
```diff
@@ -1,13 +1,14 @@
 module github.com/TwiN/gatus/v5
 
-go 1.22.2
+go 1.22.4
```
Comment (TwiN) [inline go.mod:3]:
```suggestion
go 1.22.2
```

File: config.yaml [hunk 11]
```diff
@@ -1,3 +1,4 @@
+log-level: INFO
```
Comment (TwiN) [inline config.yaml:1]:
```suggestion
```

File: watchdog/watchdog.go [hunk 12]
```diff
@@ -13,6 +13,8 @@ import (
 	"github.com/TwiN/gatus/v5/config/maintenance"
 	"github.com/TwiN/gatus/v5/metrics"
 	"github.com/TwiN/gatus/v5/storage/store"
+
```
Comment (TwiN) [inline watchdog/watchdog.go:16]:
```suggestion
```

File: watchdog/watchdog.go [hunk 13]
```diff
@@ -64,37 +66,34 @@ func execute(ep *endpoint.Endpoint, alertingConfig *alerting.Config, maintenance
...
 	if connectivityConfig != nil && connectivityConfig.Checker != nil && !connectivityConfig.Checker.IsConnected() {
-		log.Println("[watchdog.execute] No connectivity; skipping execution")
+		logr.Infof("[watchdog.execute] No connectivity; skipping execution")
 		return
 	}
-	if debug {
-		log.Printf("[watchdog.execute] Monitoring group=%s; endpoint=%s", ep.Group, ep.Name)
-	}
+	logr.Debugf("[watchdog.execute] Monitoring group=%s; endpoint=%s", ep.Group, ep.Name)
+
```
Comment (TwiN) [inline watchdog/watchdog.go:73]:
```suggestion
```

**Review by TwiN — APPROVED** [review]
Thank you for the contribution!

**Comment by TwiN** [comment]
@garyhodgson FYI I decided to move log-level to an environment variable - see #895

**Pull request merged**